### PR TITLE
Extract MainLoop out of Browser

### LINKED ISSUE #5355: Separate Browser.mainLoop from the browser API functions
If you use EmterpreterAsync it pulls in the whole Browser object, which is about 30KB of code. The majority of this is not required by EmterpreterAsync. It would be possible to separate out Browser.mainLoop and the handful of other functions which are needed (resumeAsyncCallbacks, pauseAsyncCallbacks, etc.)

While it may be possible, I'm not sure how much work it would be or whether it would be worth it... but it doesn't hurt to post an issue suggesting it.

### PR DESCRIPTION
**Pull request by curiousdannii**
Fixes #5355

I just noticed that Asyncify mode requires MainLoop, but not because of anything inherent in Asyncify mode, but just because it needs to be aware of it. It would be great if Asyncify could handle MainLoop if it's included, but to not require it. Not too sure how to do that though. Trying a simple `typeof MainLoop !== 'undefined'`

Similarly, `MainLoop.safeSetTimeout` could be split out into an even smaller library, so that if all you use is `emscripten_sleep` it wouldn't have to include all of MainLoop? Edit: This was very easy actually.

### CODE REVIEW

**Comment by curiousdannii** [comment]
There are a few errors that I couldn't make sense of :/

**Comment by sbc100** [comment]
Nice! Thanks for working on this.

Can we maybe break this down into smaller changes so it more obvious if there are any actual changes here of it is NFC.

To start with how about splitting out just the safetimers part?  That seems like it might be separable.

File: src/library_safetimers.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * @license
+ * Copyright 2020 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+// Safe timers
+var LibrarySafeTimers = {
+  $SafeTimers__postset: 'Module["requestAnimationFrame"] = function Module_requestAnimationFrame(func) { SafeTimers.requestAnimationFrame(func) };\n',
+  $SafeTimers: {
+    // abort and pause-aware versions TODO: build main loop on top of this?
```
Comment (sbc100) [inline src/library_safetimers.js:11]: Can we take this object's 7 member as make them into top level library members?

That way a program that only needs `safeSetTimeout` would depend only on that one functuion.

Of course each of the function would need to depend on `allowAsyncCallbacks` and `queuedAsyncCallbacks` so maybe that could be done with `autoAddDeps`?

File: src/library_safetimers.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
+/**
```
Comment (sbc100) [inline src/library_safetimers.js:1]: I wonder if we can come up with a better name for this library?   I'm not really sure what "safe timers" means.  Do you know the history of this name or what it really means?

Comment (curiousdannii) [reply src/library_safetimers.js:1]: I think they're "safe" in that the callbacks won't be run (immediately) if it's paused, or if the module has aborted.

**Comment by curiousdannii** [comment]
> Nice! Thanks for working on this.
> 
> Can we maybe break this down into smaller changes so it more obvious if there are any actual changes here of it is NFC.
> 
> To start with how about splitting out just the safetimers part? That seems like it might be separable.

Sure, that makes sense.

File: src/library_safetimers.js (hunk 1 above)
Comment (curiousdannii) [reply src/library_safetimers.js:11]: Sure. Should they all be prefixed with `emscripten_`?

Comment (sbc100) [reply src/library_safetimers.js:11]: No, these are library internal so should be prefixed with `$`.

Anything without a `$` is visible as a C/C++ and we don't want that in this case I think.

The `$` is only need in the `__deps` and on the decl itself..  When you call the function you don't need the `$`.

**Comment by kripken** [comment]
> Would it be worth turning the old functions into wrappers of the new with assertion warnings?

For an internal API I think that's probably not needed.

**Pull request closed without merging**